### DISCUSSION TITLE
native dll locked when unloading AppDomain (windows)

### DISCUSSION THREAD

**Discussion by viceice** [ClearFoundry/ClearScript#394] [Q&A]
I've a process architecture which uses AppDomains to seperate functionalities. 
So one AppDomain can use ClearScript.
When the AppDomain get's unloaded the `ClearScriptV8.win-x64.dll` file seems to remain locked by windows.

Any buildin way to fix that?

My idea is to copy that file to a temp folder and use `AuxiliarySearchPath` to let clearscript load the file.

**Comment by ClearScriptLib** [comment]
Hi @viceice,

ClearScript uses [`DllImportAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute) to invoke native methods. When it does so, the .NET runtime acquires a reference to the native library, loading it if necessary. Our understanding is that it never releases that reference.

In some legacy environments – as well as deployments where the native library is in a custom location – ClearScript can manually load the native library before any native calls take place; that's where `AuxiliarySearchPath` can be useful. ClearScript then releases its reference when all `V8ScriptEngine` and `V8Runtime` instances have been disposed. Unfortunately, the .NET runtime retains its reference, so the library remains locked until process termination.

A [post](https://stackoverflow.com/a/2445558/2142539) on StackOverflow suggests that one could work around that by _double-releasing_ a manually loaded library reference, but ClearScript doesn't do that. We feel that it's a dangerous hack, and no users have asked for it until now.

Perhaps we could implement forced library release as an opt-in feature. Can you tell us more about your use case?

Thanks!

**Reply by viceice** [reply]
we have a feature to self update assemblies in an appdomain. we use shadowing to not have locked assemblies.

- so we post a package via http request to main domain and that is dispatching the request to the right project appdomain.
- the project appdomain extracts the package and overwrites the assemblies 
- the project appdomain returns an unload request and get's restarted.


so the error occured on clearscript native assembly overwrite try.
our workaround for now is to copy the assembly to a known temp path and use the auxillary search path feature of clearscript to point to that path.

**Reply by ClearScriptLib** [reply]
>our workaround for now is to copy the assembly to a known temp path and use the auxillary search path feature of clearscript to point to that path.

Is this workaround sufficient for your scenario?

**Reply by viceice** [reply]
yes, it seems to work. we are using a shared path `<cache>/ClearScript/<version>/ClearScriptV8.win-x64.dll` and set `<cache>/ClearScript/<version>` as `AuxiliarySearchPath` so all AppDomains can use the shared file when using same ClearScript version.

we're using `FileVersionInfo.GetVersionInfo("ClearScriptV8.win-x64.dll").FileVersion` to get the version 😉

**Reply by AudriusButkevicius** [reply]
Seems like unloading doesn't really work as there seems to be some strangeness with suppressing finalizers etc,
effectively end up with:

![image](https://user-images.githubusercontent.com/1144861/177034939-ac8a7919-9215-4514-a657-e5da2bb057f4.png)

**Reply by viceice** [reply]
you've probably running JavaScript code while unloading appdomain. 🤷‍♂️ it's now working fine for me, let's see when next update arrives and I need to test it live

**Reply by AudriusButkevicius** [reply]
If you take the sample in:
https://github.com/dotnet/samples/blob/main/core/tutorials/Unloading/Host/Program.cs

and apply the following diff:
```diff
diff --git a/core/tutorials/Unloading/Plugin/Plugin.csproj b/core/tutorials/Unloading/Plugin/Plugin.csproj
index 6a7bde05df..b5a8dd4d3f 100644
--- a/core/tutorials/Unloading/Plugin/Plugin.csproj
+++ b/core/tutorials/Unloading/Plugin/Plugin.csproj
@@ -2,6 +2,7 @@

   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
   </PropertyGroup>

   <ItemGroup>
@@ -12,4 +13,8 @@
     <ProjectReference Include="..\PluginDependency\PluginDependency.csproj" />
   </ItemGroup>

+  <ItemGroup>
+    <PackageReference Include="Microsoft.ClearScript" Version="7.3.1" />
+  </ItemGroup>
+
 </Project>
diff --git a/core/tutorials/Unloading/Plugin/PluginClass.cs b/core/tutorials/Unloading/Plugin/PluginClass.cs
index ddc7e208c9..9f3958a77b 100644
--- a/core/tutorials/Unloading/Plugin/PluginClass.cs
+++ b/core/tutorials/Unloading/Plugin/PluginClass.cs
@@ -1,7 +1,9 @@
 ﻿using System;
+using System.Diagnostics;
 using System.Reflection;
 using System.Runtime.InteropServices;
 using System.Runtime.Loader;
+using Microsoft.ClearScript.V8;

 namespace Plugin
 {
@@ -27,12 +29,28 @@ namespace Plugin
         private static void OnPluginUnloadingRequested(AssemblyLoadContext obj)
         {
             PluginDependency.Logger.LogMessage("Cleanup of stuff preventing unloading");
+            foreach (ProcessModule module in Process.GetCurrentProcess().Modules)
+            {
+                if (module.FileName.EndsWith("ClearScriptV8.win-x64.dll"))
+                {
+                    var freeLibraryResult = FreeLibrary(module.BaseAddress);
+                    Console.WriteLine($"Freeing {module.FileName} {freeLibraryResult}");
+                }
+            }
         }

+
+        [DllImport("kernel32", SetLastError=true)]
+        static extern bool FreeLibrary(IntPtr hModule);
+
+
         // Plugin interface methods implementation

         public string GetMessage()
         {
+            using var engine = new V8ScriptEngine();
+            engine.AddHostObject("log", new Action<object>(Console.WriteLine));
+            engine.Execute("log('hello');");
             return "Hello from the unloadable plugin";
         }

```

You'll see the following output (which confirms that it is not unloaded successfully):
```
Loading assembly C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\ClearScript.V8.dll into the HostAssemblyLoadContext
Loading assembly C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\ClearScript.Core.dll into the HostAssemblyLoadContext
Loading assembly C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\Newtonsoft.Json.dll into the HostAssemblyLoadContext
Loading assembly C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\ClearScript.V8.ICUData.dll into the HostAssemblyLoadContext
hello
Response from the plugin: GetVersion(): (1.0), GetMessage(): Hello from the unloadable plugin
Loading assembly C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\PluginDependency.dll into the HostAssemblyLoadContext
Cleanup of stuff preventing unloading
Freeing C:\code\samples\core\tutorials\Unloading\Plugin\bin\Release\netcoreapp3.1\runtimes\win-x64\native\ClearScriptV8.win-x64.dll True
Unload success: False
```

**Reply by ClearScriptLib** [reply]
Hi @AudriusButkevicius,

Thanks for posting your sample.

ClearScript's V8 resources – engines, runtimes, script objects, compiled scripts, and a few odds and ends – have a historical quirk in that they have meaningful finalization even if disposed. That makes it unsafe to unload the native library as long as any such object remains reachable or GC remains pending. That explains the access violation in the screenshot above.

As for why the domain or ALC fails to unload after using V8, we have only a theory. ClearScript relies on [Reverse P/Invoke](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate) to allow the native library to call managed code. Each function pointer created for this purpose jumps to a _thunk_ that enters the appropriate context before invoking the managed code. It's likely that these thunks hold strong references that make unloading the context impossible.

Cheers!

**Reply by AudriusButkevicius** [reply]
Do you have any pointers where you think that is happening? I managed to compile the project and poke at a few places (removing "destructors" as they previously caused issues for me), but to no avail.

Are there any tools to tell me what the GC thinks is still being used?

**Reply by ClearScriptLib** [reply]
@AudriusButkevicius The method [`V8SplitProxyManaged.CreateMethodTable`](https://github.com/microsoft/ClearScript/blob/f764372dcd53dd6e8c50ae37afc78f138e4cd568/ClearScript/V8/SplitProxy/V8SplitProxyManaged.cs#L310) constructs a table of native-to-managed call thunks. This method is used to initialize a static property, so it's invoked once within each context that uses V8. Currently there's no mechanism for releasing the resources it allocates. In typical single-context applications, that's not a concern at all.

**Reply by AudriusButkevicius** [reply]
You are totally on the money.

Added a hack that exposes that object as a public, added a static list to collect the GC handles, and a method that you can call to free the handles, and it unloads properly.

I can even load and unload it multiple times, as new ALCs reinitialise the statics no problem.

What would be a sensible way to expose this?

Something that gets called as part of `OnEntityHolderDestroyed`, after freeing the native library?

**Reply by AudriusButkevicius** [reply]
Seems that there is also something about raising exceptions to the host, once it's raised, seems that stuff cannot be freed.

Any pointers on this one?

**Reply by ClearScriptLib** [reply]
Hi @AudriusButkevicius,

>Something that gets called as part of `OnEntityHolderDestroyed`, after freeing the native library?

Hmm, that's a very interesting idea – automatic and safe, with no clunky new APIs required. It should probably be done _before_ unloading the native library, and it might be wise to do it asynchronously, as `OnEntityHolderDestroyed` is called from a finalizer. We'll take a closer look.

>Seems that there is also something about raising exceptions to the host, once it's raised, seems that stuff cannot be freed.

What happens in that case? Does the final `OnEntityHolderDestroyed` not get called? That is, does `splitImplCount` not transition back to zero? In any case, automatic cleanup as you've described is still an idea worth investigating.

Thank you!

**Reply by AudriusButkevicius** [reply]
The refcount does go to zero and free does get called, but there must be something related to exceptions that leaks. I checked the AddHostObjRef part where we pass down gc handles into v8, but all those seem to be freed before freeing the library.

Is there some way to get info from the GC whats preventing the cleanup?
